---
title: "Likelihood ratio test for interaction term"
author: "Nuo Xu"
output:
  html_document:
    toc: true
    df_print: paged
  html_notebook:
    toc: true
    number_sections: true
    theme: readable
---
# Setup

## Load libraries
```{r, load-packages, warning=FALSE, message=FALSE}
library(tidyverse)
library(org.Hs.eg.db)
library(DESeq2)
library(EnsDb.Hsapiens.v79)
library(GenomicFeatures)
library(clusterProfiler)
library(here)
library(tximport)
library(EnhancedVolcano)
library(ggpubr)
library(txdbmaker)
library(clusterProfiler)
library(enrichplot)
library(patchwork)
library(readxl)
library(msigdbr)
```

# Create DESeqDataSet object from txi object

Get mappings from gene id to transcript id, this step takes a while
```{r}
txdb <- makeTxDbFromGFF(paste0(Sys.getenv("GENOMIC_DATA_DIR"), "GENCODE/Homo_sapiens.GRCh38.104.gtf"))
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")
tx2gene <- bind_rows(tx2gene, data.frame(TXNAME = c("EGFP"), GENEID = c("EGFP")))
```

Get paths to salmon output files
```{r}
files <- list.files(here("data/salmon_with_eGFP/compiled_quants_egfp"), full.names = TRUE)
names(files) <- stringr::str_extract(files, "\\d+(_\\d+)?_(Input|IP)")
```

Create txi object
```{r}
txi <- tximport(files, type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE)
```

Export TPM
```{r}
txi$abundance %>%
  as_tibble(rownames = "ENSEMBL") %>%
  write_csv(here("results/tpm.csv"))
```

Create `DESeqDataSet`
```{r}
samples <- factor(str_extract(names(files), "^.*?(?=_IP|_Input)"), levels = c("4_7", "4_9", "95_3", "87", "90", "91"))
assay <- factor(str_extract(files, "IP|Input"), levels = c("Input", "IP"))
condition <- factor(ifelse(grepl("_", samples), "PTEN_KO", "WT"), levels = c("WT", "PTEN_KO"))
coldata <- data.frame(samples = samples, assay = assay, condition = condition)
dds <- DESeqDataSetFromTximport(txi, coldata, design = ~condition)
dds <- DESeq2::estimateSizeFactors(dds)
```

# Run DESeq2

After we have created the `DESeqDataSet` object, we can subset this object to samples from the IP fraction only and from the Input fraction only, and then run DESeq on these subsets respectively.

```{r}
IP_only <- dds[, colData(dds)$assay == "IP"]
design(IP_only) <- formula(~condition)
IP_only <- DESeq(IP_only)

IP_res <- results(IP_only, contrast = c("condition", "PTEN_KO", "WT"))
IP_res$symbols <- unname(symbols[rownames(IP_res)])

write.csv(IP_res, "results/IP_res.csv")
```

# Functional interpretation

## Preparation

Get disease gene lists
```{r}
epilepsy1 <- read.csv("data/annotation/Early onset or syndromic epilepsy-1.tsv", sep = "\t")[["EnsemblId.GRch38."]]
epilepsy1 <- epilepsy1[nzchar(epilepsy1)]

epilepsy2 <- read_excel("data/epilepsy_genes_mao.xlsx", sheet = 1)
epilepsy2 <- epilepsy2["Gene"] %>% unique() %>% unlist() %>% unname() 
epilepsy2 <- AnnotationDbi::select(org.Hs.eg.db, keys = epilepsy2, columns = c("ENSEMBL", "SYMBOL"), keytype = "ALIAS") %>% pull(ENSEMBL) %>% unique()

epilepsy <- union(epilepsy1, epilepsy2)

SFARI_all <- read.csv("data/annotation/SFARI-Gene_genes_03-28-2024release_05-28-2024export.csv")[["ensembl.id"]]
SFARI_all <- SFARI_all[nzchar(SFARI_all)]

gene_sets <- c(list(epilepsy = epilepsy, SFARI = SFARI_all), split(GO_df$ensembl_gene, GO_df$gs_name))
```

Get GO groups of interest
```{r}
gs_name_list_all <- msigdbr(species = "Homo sapiens", category = "C5")
gs_name_list <- pull(read_excel("data/annotation/GOBP_enrichmentterms.xlsx", col_names = FALSE), `...1`)
gs_name_list <- str_replace_all(gs_name_list, "\r\n", "")
write.csv(gs_name_list_all[match(gs_name_list, gs_name_list_all$gs_name), "gs_exact_source"], "data/annotation/octavia_GO_terms.csv", row.names = FALSE, quote = FALSE)

# Then run this list of GO terms through the Revigo web app
gs_exact_source_reduced <- read.csv("data/annotation/Revigo_BP_Octavia.tsv", sep = "\t")[["TermID"]]
GO_df <- msigdbr(species = "Homo sapiens") %>% 
    dplyr::filter(gs_exact_source %in% gs_exact_source_reduced)
GO_df$gs_name <- as.factor(GO_df$gs_name)
```

Combine disease genes and genes in GO terms into one single dataframe
```{r}
gene_sets <- c(list(epilepsy = epilepsy, SFARI = SFARI_all), split(GO_df$ensembl_gene, GO_df$gs_name))
```

Get upregulated genes in IP fraction
```{r}
genes_of_interest_IP <- IP_res %>%
    dplyr::filter(padj < 0.05) %>% 
    row.names()
temp <- lapply(gene_sets, function(x) {genes_of_interest_IP %in% x}) %>% data.frame()
temp$ENSEMBL <- genes_of_interest_IP
```

## Figure. Top 30 upregulated genes that are also in the epilepsy database
```{r}
mapping <- AnnotationDbi::select(org.Hs.eg.db, keys = temp$ENSEMBL, columns = c("ENSEMBL", "SYMBOL"), keytype = "ENSEMBL")
mapping <- mapping %>% 
    group_by(ENSEMBL) %>%
    slice(1) %>%
    ungroup()

IP_res$ENSEMBL <- rownames(IP_res)

temp <- temp %>% 
    merge(IP_res, by = "ENSEMBL") %>%
    merge(mapping, by = "ENSEMBL") %>%
    dplyr::select(c("ENSEMBL", "SYMBOL", "epilepsy", "SFARI", "log2FoldChange", "padj")) %>% 
    mutate(score = epilepsy + SFARI) %>% 
    arrange(desc(score), desc(abs(log2FoldChange)))
```

```{r}
my_theme <- theme_bw() +
    theme(
        axis.title.x = element_text(size = 5),
        axis.title.y = element_text(size = 5),
        axis.text.x = element_text(size = 5),
        axis.text.y = element_text(size = 5),
    )

theme_set(my_theme)

temp %>% 
    dplyr::filter(epilepsy == TRUE) %>% 
    arrange(desc(log2FoldChange)) %>% 
    slice(1:30) %>% 
    ggbarplot(x = "SYMBOL", y = "log2FoldChange", x.text.angle = 90) +
    xlab("Gene symbols") +
    ylab("Log2 fold change") +
    ggtitle("Top 30 upreguated genes that are also in the epilepsy database")
ggsave("results/figures/top_30_epilepsy.pdf", width = 6, height = 4)
```

## Enrichment analysis
```{r prepare data for enrichment analysis}
geneList <- IP_res %>% 
  as_tibble() %>%
  dplyr::filter(padj < 0.05 & log2FoldChange > 0.2) %>%
  pull(ENSEMBL)

background_genes <- IP_res %>% 
  as_tibble() %>%
  pull(ENSEMBL)
```

```{r create-enrichGO objects}
ego_MF <- enrichGO(
  gene = geneList, 
  universe = background_genes,
  keyType = "ENSEMBL",
  OrgDb = org.Hs.eg.db, 
  pAdjustMethod = "BH", 
  qvalueCutoff = 0.05, 
  readable = TRUE,
  ont = "MF"
  )

ego_BP <- enrichGO(
  gene = geneList, 
  universe = background_genes,
  keyType = "ENSEMBL",
  OrgDb = org.Hs.eg.db, 
  pAdjustMethod = "BH", 
  qvalueCutoff = 0.05, 
  readable = TRUE,
  ont = "BP"
)

ego_CC <- enrichGO(
  gene = geneList, 
  universe = background_genes,
  keyType = "ENSEMBL",
  OrgDb = org.Hs.eg.db, 
  pAdjustMethod = "BH", 
  qvalueCutoff = 0.05, 
  readable = TRUE,
  ont = "CC"
)
```

```{r dotplot}
width <- 5.5
height <- 4

MF_dotplot <- dotplot(ego_MF, showCategory=10)
ggsave("figures/clusterProfiler/IP_MF_GO_enrichment.pdf", width = width, height = height)

BP_dotplot <- dotplot(ego_BP, showCategory=10)
ggsave("figures/clusterProfiler/IP_BP_GO_enrichment.pdf", width = width, height = height)

CC_dotplot <- dotplot(ego_CC, showCategory=10)
ggsave("figures/clusterProfiler/IP_CC_GO_enrichment.pdf", width = width, height = height)
```

```{r emap}
width <- 7
height <- 5.5

MF_emap <- pairwise_termsim(ego_MF) %>% 
  emapplot(showCategory=20)
ggsave("figures/clusterProfiler/IP_MF_GO_emap.pdf", width = width, height = height)

BP_emap <- pairwise_termsim(ego_BP) %>% 
  emapplot(showCategory=20)
ggsave("figures/clusterProfiler/IP_BP_GO_emap.pdf", width = width, height = height)

CC_emap <- pairwise_termsim(ego_CC) %>% 
  emapplot(showCategory=20)
ggsave("figures/clusterProfiler/IP_CC_GO_emap.pdf", width = width, height = height)
```

```{r combining-figures}
CC_dotplot + CC_emap + MF_dotplot + MF_emap + BP_dotplot + BP_emap +
  plot_annotation(tag_levels = 'A') + plot_layout(ncol = 2, widths = c(1, 2))
ggsave("figures/clusterProfiler/IP_GO_enrichment_combined_2col.pdf", width = 11, height = 16)
```

# Volcano plots
```{r}
IP_res$volcano_label <- ifelse(IP_res$symbols == "KCNA1", "KCNA1", NA)
EnhancedVolcano(
  IP_res,
  lab = IP_res$volcano_label,
  x = "log2FoldChange",
  y = "padj",
  title = "IP fraction",
  pCutoff = 0.05,
  xlim = c(-10, 10),
  ylim = c(0, 40),
  drawConnectors = TRUE, widthConnectors = 0.75
)
ggsave("figures/IP_volcano.pdf", width = 5, height = 5.5)
```