---
title: "Likelihood ratio test for interaction term"
author: "Nuo Xu"
output:
  html_document:
    toc: true
    df_print: paged
  html_notebook:
    toc: true
    number_sections: true
    theme: readable
---
# Setup

## Load libraries
```{r, load-packages, warning=FALSE, message=FALSE}
library(tidyverse)
library(org.Hs.eg.db)
library(DESeq2)
library(EnsDb.Hsapiens.v79)
library(GenomicFeatures)
library(clusterProfiler)
library(ggVennDiagram)
library(here)
library(tximport)
```

## Load data and helper functions

```{r}
symbols <- mapIds(
  EnsDb.Hsapiens.v79,
  keys = keys(EnsDb.Hsapiens.v79, "GENEID"),
  column = c("SYMBOL"),
  keytype = "GENEID"
)

gene_types <- mapIds(
  EnsDb.Hsapiens.v79,
  keys = keys(EnsDb.Hsapiens.v79, "GENEID"),
  column = c("GENEBIOTYPE"),
  keytype = "GENEID"
)

# Import helper functions that we'll use later
source(here("src", "helper_functions.R"))
```

# Create DESeqDataSet object from txi object

Get mappings from gene id to transcript id, this step takes a while
```{r}
txdb <- makeTxDbFromGFF(here("data/Homo_sapiens.GRCh38.104.gtf"))
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")
tx2gene <- bind_rows(tx2gene, data.frame(TXNAME = c("EGFP"), GENEID = c("EGFP")))
```

Get paths to salmon output files
```{r}
files <- list.files(here("data/salmon_with_eGFP/compiled_quants_egfp"), full.names = TRUE)
names(files) <- stringr::str_extract(files, "\\d+(_\\d+)?_(Input|IP)")
```

Create txi object
```{r}
txi <- tximport(files, type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE)
```

Create column metadata for `DESeqDataSet`
```{r}
samples <- factor(str_extract(names(files), "^.*?(?=_IP|_Input)"), levels = c("4_7", "4_9", "95_3", "87", "90", "91"))
assay <- factor(str_extract(files, "IP|Input"), levels = c("Input", "IP"))
# Note here that the last level of the condition factor is used as the numerator
condition <- factor(ifelse(grepl("_", samples), "PTEN_KO", "WT"), levels = c("WT", "PTEN_KO"))
coldata <- data.frame(samples = samples, assay = assay, condition = condition)
```

Create `DESeqDataSet`
```{r}
dds <- DESeqDataSetFromTximport(txi, coldata, design = ~ condition)
dds <- DESeq2::estimateSizeFactors(dds)
```

# Run DESeq2

After we have created the `DESeqDataSet` object, we can subset this object to samples from the IP fraction only and from the Input fraction only, and then run DESeq on these subsets respectively.

## IP samples only

```{r}
IP_only <- dds[, colData(dds)$assay == "IP"]
IP_only <- DESeq(IP_only)
results(IP_only, contrast=c("condition", "WT", "PTEN_KO"))
```

## Input samples only

```{r}
Input_only <- dds[, colData(dds)$assay == "Input"]
design(Input_only) <- formula(~ condition)
Input_only <- DESeq(Input_only)
results(IP_only, contrast=c("condition", "WT", "PTEN_KO"))
```

## Interaction term

```{r}
# Likelihood ratio test
design(dds) <- formula(~ assay + condition + condition:assay)
dds_interaction <- DESeq(dds, test = "LRT", reduced = ~ assay + condition)
results(dds_interaction)
```
# Focusing on interaction term results

```{r}
# The original log fold changes
dds_interaction %>%
  results() %>% 
  as_tibble(rownames = "ENSEMBL") %>% 
  mutate(symbol = symbols[rownames(dds_interaction)]) %>% 
  mutate(gene_type = gene_types[rownames(dds_interaction)]) %>% 
  write_csv(here("results/assayIP_conditionPTEN_KO_raw.csv"))

# Shrink log fold changes, which is useful for ranking and visualization
res_lfcShrink <- dds_interaction %>%
  lfcShrink(coef="assayIP.conditionPTEN_KO", type="apeglm") %>%
  as_tibble(rownames = "ENSEMBL") %>% 
  mutate(symbol = symbols[rownames(dds_interaction)]) %>% 
  mutate(gene_type = gene_types[rownames(dds_interaction)])

res_lfcShrink %>% write_csv(here("results/assayIP_conditionPTEN_KO.csv"))
```

## Individual gene counts
You can visualize the distribution of individual genes using this function
```{r}
plotCounts(dds_interaction, gene="ENSG00000204789", intgroup=c("assay", "condition"))
```

## Functional interpretation

### Over-representation analysis

For over-representation analysis, we first rank all genes by log2FoldChange and then run ORA on the top 10% and the bottom 10% of the gene list separately. 
```{r}
ego_for_plotting <- get_ego_for_plotting(res_lfcShrink)
plot_ego(ego_for_plotting)
```

### Gene set enrichment analysis

Here we are using the `log2FoldChange * -log10(pvalue)` as the ranking metric for the gene set enrichment analysis
```{r}
# Calculate the ranking metric
res_lfcShrink <- res_lfcShrink %>% 
  dplyr::filter(complete.cases(pvalue) & complete.cases(log2FoldChange)) %>% 
  mutate(score = log2FoldChange * -log10(pvalue)) %>% 
  arrange(desc(score))

# Order the genes in terms of the ranking metric
geneList <- res_lfcShrink %>% 
  mutate(log10pvalue = (-log10(pvalue))) %>% 
  arrange(desc(log10pvalue)) %>% 
  pull(log10pvalue)
names(geneList) <- res_lfcShrink %>% 
  mutate(log10pvalue = (-log10(pvalue))) %>% 
  arrange(desc(log10pvalue)) %>% 
  pull(ENSEMBL)
```

```{r}
# Run gene set enrichment analysis and visualize
ego3 <- gseGO(
  geneList = geneList,
  OrgDb = org.Hs.eg.db,
  ont = "ALL",
  keyType = "ENSEMBL",
  minGSSize = 100,
  maxGSSize = 500,
  pvalueCutoff = 0.05,
  verbose = FALSE)

dotplot(ego3)
```

## SFARI genes

```{r}
# get the list of SFARI genes
SFARI_genes <- read_csv(here("data/SFARI-Gene_genes_03-28-2024release_05-28-2024export.csv"))

# filter for genes with padj < 0.2 and also present in the SFARI gene list
SFARI_plotting <- res_lfcShrink %>% 
  filter(ENSEMBL %in% SFARI_genes[["ensembl-id"]]) %>% 
  filter(padj < 0.2) %>% 
  left_join(SFARI_genes, join_by(ENSEMBL == `ensembl-id`))
```

```{r}
SFARI_plotting %>%
  mutate(`gene-score` = if_else(is.na(`gene-score`), 3, `gene-score`)) %>% 
  ggplot(aes(x = reorder(symbol, log2FoldChange), y = log2FoldChange, fill = -log10(pvalue))) +
    geom_bar(stat = "identity") +
    scale_fill_gradient() +
    theme(
      axis.text.x = element_text(
        angle = 90, vjust = 0.5, hjust=1, 
        color=ifelse(SFARI_plotting[["gene-score"]] == 1,"red", "black")),
      axis.title.x = element_blank()
      ) +
    ggtitle("Log2 fold changes of ASD-risk genes")
```


