---
title: "DESeq calculation of RE and shifts in RE"
output:
  html_notebook:
    toc: true
    number_sections: true
    theme: readable
author: Nuo Xu
---
# Preprocessing

## Loading libraries

```{r load-libraries}
library(ggplot2)
library(scales)
library(dplyr)
library(tidyr)
library(DESeq2)
library(stringr)
library(patchwork)
```

## Obtain gene symbols
```{r}
# Obtain gene symbols
gtf_path <- "/scratch/s/shreejoy/nxu/Genomic_references/hg38/Raw/Homo_sapiens.GRCh38.111.gtf"
annotation_from_gtf <- rtracklayer::import(gtf_path) %>%
    dplyr::as_tibble() %>%
    dplyr::filter(type == "gene") %>%
    dplyr::select(gene_id, gene_name, gene_biotype)
```

## Define helper functions
```{r}
get_RE <- function(txi, min_row_sum, clean = FALSE) {
    filtered_counts <- txi$counts[rowSums(txi$counts) > min_row_sum, ]
    filtered_counts <- filtered_counts %>%
        as.data.frame() %>%
        mutate(across(everything(), as.integer))
    samples <- factor(
        str_extract(colnames(filtered_counts), "4_7|4_9|95_3|87|90|91"),
        levels = c("4_7", "4_9", "95_3", "87", "90", "91"))
    assay <- factor(str_extract(colnames(filtered_counts), "IP|Input"), levels = c("IP", "Input"))
    condition <- factor(ifelse(grepl("_", samples), "PTEN_KO", "WT"), levels = c("WT", "PTEN_KO"))
    coldata <- data.frame(samples = samples, assay = assay, condition = condition)
    dds <- DESeq2::DESeqDataSetFromMatrix(
        countData = filtered_counts,
        colData = coldata,
        design = ~samples
    )
    dds <- DESeq2::estimateSizeFactors(dds)
    up_vec <- counts(dds, normalized = TRUE)[, colData(dds)$assay == "IP"]
    down_vec <- counts(dds, normalized = TRUE)[, colData(dds)$assay == "Input"]
    RE <- as.data.frame(up_vec / down_vec)
    if (clean == TRUE) {
        RE[!apply(RE, 1, function(x) any(is.infinite(x) | is.nan(x))), ]
    } else if (clean == FALSE) {
        RE
    }
}

get_RE_normalized <- function(dds, min_row_sum, clean = FALSE) {
    filtered_counts <- txi$counts[rowSums(txi$counts) > min_row_sum, ]
    filtered_counts <- filtered_counts %>%
        as.data.frame() %>%
        mutate(across(everything(), as.integer))
    samples <- factor(
        str_extract(colnames(filtered_counts), "4_7|4_9|95_3|87|90|91"),
        levels = c("4_7", "4_9", "95_3", "87", "90", "91"))
    assay <- factor(str_extract(colnames(filtered_counts), "IP|Input"), levels = c("IP", "Input"))
    condition <- factor(ifelse(grepl("_", samples), "PTEN_KO", "WT"), levels = c("WT", "PTEN_KO"))
    coldata <- data.frame(samples = samples, assay = assay, condition = condition)
    dds <- DESeq2::DESeqDataSetFromMatrix(
        countData = filtered_counts,
        colData = coldata,
        design = ~samples
    )
    dds <- DESeq2::estimateSizeFactors(dds)
    up_vec <- counts(dds, normalized = TRUE)[, colData(dds)$assay == "IP"]
    down_vec <- counts(dds, normalized = TRUE)[, colData(dds)$assay == "Input"]
    RE <- as.data.frame(up_vec / down_vec)
    if (clean == TRUE) {
        RE <- RE[!apply(RE, 1, function(x) any(is.infinite(x) | is.nan(x))), ]
    }
    as.data.frame(preprocessCore::normalize.quantiles(RE, keep.names = TRUE))
}
```

## Loading data
```{r}
txi_nuo <- readRDS("proc/txi_nuo.rds")
txi_wendy <- readRDS("proc/txi_wendy.rds")
```

# Comparing results

```{r}
RE <- get_RE(txi_wendy, 150, FALSE)
RE_normalized <- get_RE_normalized(txi_wendy, 150, FALSE)
```

# Visualizations

## Distribution of raw counts
```{r}
data.frame(count = rowSums(counts(dds_wendy))) %>%
    ggplot(aes(x = count)) +
    geom_histogram(bins = 100) +
    geom_vline(aes(xintercept = 150), color = "red", linetype = "dashed", size = 1) +
    scale_x_continuous(limits = c(0, 500)) +
    scale_y_continuous(limits = c(0, 500))

sum(rowSums(counts(dds_wendy)) > 150) / dim(dds_wendy)[1]
```

## Reproducing figures from the protocol
```{r}
# before normalization boxplot
b1 <- wendy_RE %>%
    mutate(across(everything(), function(x) {
        log(x) + 1
    })) %>%
    pivot_longer(cols = everything(), names_to = "sample", values_to = "RE") %>%
    mutate(sample = factor(sample, levels = c("4_7", "4_9", "95_3", "87", "90", "91"))) %>%
    mutate(condition = ifelse(str_detect(sample, "_"), "KO", "WT")) %>%
    ggplot(aes(x = sample, y = RE, colour = condition)) +
    geom_boxplot(outliers = FALSE) +
    theme(
        axis.title.x = element_blank(),
        axis.text.x = element_blank()
    )

# after normalization boxplot
b2 <- wendy_RE_normalized %>%
    mutate(across(everything(), function(x) {
        log(x) + 1
    })) %>%
    pivot_longer(cols = everything(), names_to = "sample", values_to = "RE") %>%
    mutate(sample = factor(sample, levels = c("4_7", "4_9", "95_3", "87", "90", "91"))) %>%
    mutate(condition = ifelse(str_detect(sample, "_"), "KO", "WT")) %>%
    ggplot(aes(x = sample, y = RE, colour = condition)) +
    geom_boxplot(outliers = FALSE)

# WT 3 time points before normalization
p1 <- wendy_RE %>%
    filter_if(is.numeric, all_vars(. < 100 | is.na(.))) %>%
    mutate(KO = rowMeans(dplyr::select(., contains(c("4_7", "4_9", "95_3"))), na.rm = TRUE)) %>%
    mutate(WT = rowMeans(dplyr::select(., contains(c("87", "90", "91"))), na.rm = TRUE)) %>%
    ggplot(aes(WT, KO)) +
    geom_bin_2d(bins = 80) +
    geom_abline(intercept = 0, slope = 1, color = "red") +
    geom_smooth(color = "red", linetype = 2) +
    scale_x_continuous(limits = c(0, 5)) +
    scale_y_continuous(limits = c(0, 5)) +
    ggtitle("before normalization")

# WT 3 time points after normalization
p2 <- wendy_RE_normalized %>%
    filter_if(is.numeric, all_vars(. < 100 | is.na(.))) %>%
    mutate(KO = rowMeans(dplyr::select(., contains(c("4_7", "4_9", "95_3"))), na.rm = TRUE)) %>%
    mutate(WT = rowMeans(dplyr::select(., contains(c("87", "90", "91"))), na.rm = TRUE)) %>%
    ggplot(aes(WT, KO)) +
    geom_bin2d(bins = 80) +
    geom_abline(intercept = 0, slope = 1, color = "red") +
    geom_smooth(color = "red", linetype = 2) +
    scale_x_continuous(limits = c(0, 5)) +
    scale_y_continuous(limits = c(0, 5)) +
    ggtitle("after normalization")

(b1 / b2) | (p1 / p2)
ggsave("/scratch/s/shreejoy/nxu/trap_seq/results/figures/protocol_Figure_2_octavia.png", width = 10, height = 8)
```

## Reproducing figures from Rodriguez et al 2020.

### Figure 1D

Get TPM from txi object
```{r}
txi <- readRDS("proc/txi_wendy.rds")
tpm <- txi$abundance %>% as.data.frame()

KO_tpm <- tpm %>%
    dplyr::select(contains(c("4_7", "4_9", "95_3"))) %>%
    mutate(across(everything(), function(x) log10(x))) %>%
    filter(rowSums(dplyr::select(., contains("Input"))) > -2) %>%
    filter(rowSums(dplyr::select(., contains("IP"))) > -2) %>%
    mutate(IP = rowMeans(dplyr::select(., contains("IP")), na.rm = TRUE)) %>%
    mutate(Input = rowMeans(dplyr::select(., contains("Input")), na.rm = TRUE)) %>%
    dplyr::select(c("IP", "Input")) %>%
    mutate(condition = "KO")

WT_tpm <- tpm %>%
    dplyr::select(contains(c("87", "90", "91"))) %>%
    mutate(across(everything(), function(x) log10(x))) %>%
    filter(rowSums(dplyr::select(., contains("Input"))) > -2) %>%
    filter(rowSums(dplyr::select(., contains("IP"))) > -2) %>%
    mutate(IP = rowMeans(dplyr::select(., contains("IP")), na.rm = TRUE)) %>%
    mutate(Input = rowMeans(dplyr::select(., contains("Input")), na.rm = TRUE)) %>%
    dplyr::select(c("IP", "Input")) %>%
    mutate(condition = "WT")

rbind(KO_tpm, WT_tpm) %>%
    ggplot(aes(x = Input, y = IP)) +
    geom_point(alpha = 0.5) +
    geom_abline(intercept = 0, slope = 1, linetype = "dashed") +
    facet_grid(cols = vars(condition)) +
    xlab("Input RNA-seq (log10 TPM)") +
    ylab("IP RNA-seq (log10 TPM)")

ggsave("results/figures/Figure_1D.png")
```   

### Plotting distribution of eGFP counts in IP vs input samples
```{r}
tpm["EGFP", ] %>%
    t() %>%
    as.data.frame() %>%
    mutate(sample = str_extract(row.names(.), "IP|Input")) %>%
    ggplot(aes(x = sample, y = EGFP)) +
    geom_boxplot() +
    geom_jitter(width = 0.2, alpha = 0.5)
ggsave("results/figures/RE_distribution.png")
```