---
title: "DEG analysis"
output:
  html_notebook:
    toc: true
    number_sections: true
    theme: readable
author: Nuo Xu
---
# Introduction
In this RMarkdown, we'll explore some other methods for finding differentially translated genes
in the PTEN KO vs WT samples. 

# Preparation

## Loading libraries

```{r load-libraries, message=FALSE, warning=FALSE, echo=FALSE}
library(ggplot2)
library(here)
library(dplyr)
library(tidyr)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(org.Hs.eg.db)
library(EnhancedVolcano)
library(limma)
```

## Create DESeqDataSet object

```{r}
txdb <- makeTxDbFromGFF("/scratch/s/shreejoy/nxu/Genomic_references/hg38/Raw/Homo_sapiens.GRCh38.104.gtf")
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")
tx2gene <- bind_rows(tx2gene, data.frame(TXNAME = c("EGFP"), GENEID = c("EGFP")))
files <- list.files("salmon_with_eGFP/compiled_quants_egfp", full.names = TRUE)
names(files) <- stringr::str_extract(files, "(?<=/)[^/]+(?=_S)")
txi <- tximport(files, type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE)

samples <- names(files)
assay <- str_extract(files, "IP|Input")
condition <- ifelse(grepl("_", samples), "PTEN_KO", "WT")
coldata <- data.frame(samples = samples, assay = assay, condition = condition)
dds <- DESeqDataSetFromTximport(txi, coldata, design = ~ samples)
dds <- DESeq2::estimateSizeFactors(dds)
saveRDS(dds, "proc/dds.rds")
```

## Load DESeqDataSet and gene symbols
```{r}
dds <- readRDS("proc/dds.rds")
symbols <- mapIds(org.Hs.eg.db, keys = keys(org.Hs.eg.db, "ENSEMBL"),
  column = c('SYMBOL'), keytype = 'ENSEMBL')
```

# Exploratory analysis and visualization

```{r pre-filtering}
smallestGroupSize <- 4
keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
dds <- dds[keep, ]
```

Variance-stabalizing transformation
```{r rlg}
rld <- rlog(dds, blind = FALSE)
```

Sample distance heatmap
```{r sampleDistMatrix}
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$assay, rld$condition, rld$samples, sep = " - ")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
```

PCA plot
```{r pca}
plotPCA(rld, intgroup = c("assay", "condition"))
```

MDS plot
```{r mds}
mds <- as.data.frame(colData(rld))  %>%
         cbind(cmdscale(sampleDistMatrix))
ggplot(mds, aes(x = `1`, y = `2`, color = assay, shape = condition)) +
  geom_point(size = 3) + coord_fixed() + ggtitle("MDS with rlg data")
```

Plot counts of specific genes for inspection
```{r plotCounts}
topGene <- rownames(res)[which.min(res$padj)]
plotCounts(dds, gene = "ENSG00000174469", intgroup=c("condition"))
```

# The interaction term approach in DESeq2

## Running DESeq
```{r results}
colData(dds)$assay <- factor(colData(dds)$assay)
colData(dds)$condition <- factor(colData(dds)$condition)
design(dds) <- formula(~ assay + condition + assay:condition)
dds <- DESeq(dds, test="LRT", reduced = ~ assay + condition)
res <- results(dds)

# getting normalized counts
counts(dds, normalized = TRUE) %>% as.data.frame() %>% write.csv("proc/normalized_counts.csv") 
```

## Eploratory figures

MA plot
```{r plotMA}
DESeq2::plotMA(res)
```

Getting gene symbols
```{r}
indices <- match(rownames(res), names(symbols))
res$SYMBOL <- ifelse(is.na(indices), rownames(res), symbols[indices])
```

```{r volcano-plot}
EnhancedVolcano(
  res,
  lab = res$SYMBOL,
  x = 'log2FoldChange',
  y = 'pvalue',
  pCutoff = 0.1,
  title = 'Interaction approach in DESeq2'
  )
```

```{r rank-genes}
resLFC <- lfcShrink(dds, coef="condition_WT_vs_PTEN_KO", type="apeglm")
resLFC %>%
  as.data.frame() %>%
  mutate(SYMBOL = symbols[rownames(.)]) %>% 
  # dplyr::filter(padj < 0.05) %>%
  arrange(desc(log2FoldChange)) %>% 
  write.csv("proc/interaction_term_results.csv")
```

```{r plotCounts}
plotCounts(dds, gene = "ENSG00000183873", intgroup = c("assay", "condition"), main = "SCN5A")
plotCounts(dds, gene = "ENSG00000134597", intgroup = c("assay", "condition"), main = "RBMX")
plotCounts(dds, gene = "ENSG00000174469", intgroup = c("assay", "condition"), main = "CNTNAP2")
```

```{r GSEA}
library(clusterProfiler)

genes <- resLFC %>% 
  as.data.frame() %>%
  dplyr::filter(padj < 0.05) %>% 
  dplyr::select(log2FoldChange)

geneList <- as.vector(genes$log2FoldChange)
names(geneList) <- rownames(genes)
geneList <- sort(geneList, decreasing = TRUE)

ggo <- gseGO(gene     = geneList,
               OrgDb    = org.Hs.eg.db,
               keyType  = "ENSEMBL",
               ont      = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05)

dotplot(ggo, showCategory=30) + ggtitle("dotplot for GSEA")
```

## Replicate Figure 2A in Rodriguez et al. 2020
```{r}
SFARI_genes <- read.csv("data/SFARI-Gene_genes_03-28-2024release_04-23-2024export.csv")
# SFARI_genes <- SFARI_genes %>% filter(gene.score %in% c(1, 2))
SFARI_genes <- SFARI_genes %>% select(c("ensembl.id", "gene.score", "gene.symbol"))

SFARI_res <- res %>%
  as.data.frame() %>%
  mutate(ensembl.id = rownames(.)) %>%
  inner_join(SFARI_genes, join_by("ensembl.id"))

SFARI_res %>%
  filter(padj < 0.05 & abs(log2FoldChange) > 0.2) %>%
  mutate(padj = -log10(padj)) %>%
  ggplot(aes(x = reorder(gene.symbol, -log2FoldChange), y = log2FoldChange, fill = padj)) +
    geom_bar(stat = "identity") +
    scale_fill_continuous(low = "white", high = "#a01ef1") +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
      axis.title.x = element_blank()
      ) +
    ggtitle("log2(RE FC) between PTEN KO and WT")
ggsave("results/figures/SFARI_genes.png", width = 10, height = 5, dpi = 600)
```

# Differential gene expression for each assay respectively

## Run DESeq on Input sample

```{r DESeq}
Input <- dds[colData(dds)$assay == "Input", ]
colData(Input)$condition <- factor(colData(Input)$condition, levels = c("PTEN_KO", "WT"))
design(Input) <- formula(~ condition)
Input <- DESeq(Input)
res_input <- results(Input, contrast = c("condition", "PTEN_KO", "WT"))
```

```{r}
DESeq2::plotMA(res_input)
```

```{r volcano-plot}
EnhancedVolcano(
  res_input,
  lab = res_input$SYMBOL,
  x = 'log2FoldChange',
  y = 'pvalue',
  title = 'Translationally regulated in input samples'
  )
```

## Run DESeq on IP sample

```{r DESEq}
IP <- dds[colData(dds)$assay == "IP", ]
colData(IP)$condition <- factor(colData(IP)$condition, levels = c("PTEN_KO", "WT"))
design(IP) <- formula(~ condition)
IP <- DESeq(IP)
res_ip <- results(IP, contrast = c("condition", "PTEN_KO", "WT"))

res_ip <- res_ip %>% 
  as.data.frame() %>% 
  mutate(ENSEMBL = rownames(.)) %>%
  left_join(annots, join_by("ENSEMBL"))
```

```{r}
plotMA(res_ip)
```

```{r volcano-plot}
EnhancedVolcano(res_ip,
  lab = res_ip$SYMBOL,
  x = 'log2FoldChange',
  y = 'padj',
  title = 'Translationally regulated in IP samples',
  pCutoff = 0.1
  )
```