---
title: "DEG analysis"
output:
  html_notebook:
    toc: true
    number_sections: true
    theme: readable
author: Nuo Xu
---
# Introduction
In this RMarkdown, we'll explore the interaction term method as described in 
[DESeq2 testing ratio of ratios (RIP-Seq, CLIP-Seq, ribosomal profiling)](https://support.bioconductor.org/p/61509/)

# Preparation

## Loading libraries

```{r load-libraries, message=FALSE, warning=FALSE, echo=FALSE}
library(ggplot2)
library(here)
library(dplyr)
library(tximport)
library(GenomicFeatures)
library(tidyr)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(org.Hs.eg.db)
library(EnhancedVolcano)
library(stringr)
```

## Create DESeqDataSet object

```{r}
txdb <- makeTxDbFromGFF("/scratch/s/shreejoy/nxu/Genomic_references/hg38/Raw/Homo_sapiens.GRCh38.104.gtf")
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")
tx2gene <- bind_rows(tx2gene, data.frame(TXNAME = c("EGFP"), GENEID = c("EGFP")))
files <- list.files("salmon_with_eGFP/compiled_quants_egfp", full.names = TRUE)
names(files) <- stringr::str_extract(files, "\\d+(?:_\\d+)?")
txi <- tximport(files, type = "salmon", tx2gene = tx2gene, ignoreTxVersion = TRUE)

samples <- names(files)
assay <- str_extract(files, "IP|Input")
assay <- factor(assay, levels = c("IP", "Input"))
condition <- ifelse(grepl("_", samples), "PTEN_KO", "WT")
condition <- factor(condition, levels = c("PTEN_KO", "WT"))
coldata <- data.frame(samples = samples, assay = assay, condition = condition)
dds <- DESeqDataSetFromTximport(txi, coldata, design = ~ assay + condition + assay:condition)
dds <- DESeq2::estimateSizeFactors(dds)

# save DESeqDataSet
saveRDS(dds, "proc/dds.rds")

# save normalized counts
counts(dds, normalized = TRUE) %>% as.data.frame() %>% write.csv("proc/normalized_counts.csv")
```

## Load DESeqDataSet and gene symbols
```{r}
dds <- readRDS("proc/dds.rds")
symbols <- mapIds(org.Hs.eg.db, keys = keys(org.Hs.eg.db, "ENSEMBL"),
  column = c('SYMBOL'), keytype = 'ENSEMBL')
```

# The interaction term approach in DESeq2

## Running DESeq
```{r results}
design(dds) <- formula(~ assay + condition + assay:condition)
dds <- DESeq(dds, test="LRT", reduced = ~ assay + condition)
res <- results(dds)
```

## Visualizations

Getting gene symbols
```{r}
indices <- match(rownames(res), names(symbols))
res$SYMBOL <- ifelse(is.na(indices), rownames(res), symbols[indices])
```

Volcano plot
```{r volcano-plot}
EnhancedVolcano(
  res,
  lab = res$SYMBOL,
  x = 'log2FoldChange',
  y = 'pvalue',
  pCutoff = 0.1,
  title = 'Interaction approach in DESeq2'
  )
```

## Get gene rankings
```{r compare}
normalized_counts <- counts(dds, normalized = TRUE)
RE <- normalized_counts[, colData(dds)$assay == "IP"] / normalized_counts[, colData(dds)$assay == "Input"]
up_vector <- RE[, str_detect(colnames(RE), "_")] %>% rowMeans(na.rm = TRUE)
down_vector <- RE[, !str_detect(colnames(RE), "_")] %>% rowMeans(na.rm = TRUE)
fcRE <- up_vector / down_vector
lfcRE <- log2(fcRE)

res$lfcRE <- lfcRE[rownames(res)]
res <- res %>% as.data.frame()
res %>% filter(SYMBOL == "CNTNAP2")
res %>% filter(SYMBOL == "CHRNA7")
```

```{r}
res %>% as.data.frame() %>% write.csv("proc/interaction_term_results.csv")
```

```{r rank-genes}
# resLFC <- lfcShrink(dds, contrast="assayInput.conditionWT")
# resLFC %>%
#   as.data.frame() %>%
#   mutate(SYMBOL = symbols[rownames(.)]) %>%
#   # dplyr::filter(padj < 0.05) %>%
#   arrange(desc(log2FoldChange)) %>%
#   write.csv("proc/interaction_term_results.csv")
```